### PR TITLE
Add support for SSL context and SNI

### PR DESCRIPTION
**Pull request by Nzbuu**
Use SSLContext objects, if available, since this provides more options and features including SNI (#11). Also, allows default system context from `ssl.create_default_context()` to be used, improves encapsulation and supports dependency injection. 

**Update**: Removed support for SSL without SSLContext (#115)

### CODE REVIEW

**Comment by Nzbuu** [comment]
I've rebased this on the latest `develop` branch to get the updated tests, etc.

**Comment by anthonyrisinger** [comment]
How can we move this forward? Is more work or review needed? SNI is the greatest!

**Comment by PierreF** [comment]
Just spotted a small unexpected behavious: tls_insecure_set() does not work when using tls_set_context() with a ssl.create_default_context().

This is caused by [SSLContext.check_hostname](https://docs.python.org/2.7/library/ssl.html#ssl.SSLContext.check_hostname) that already performs the [match_hostname](https://docs.python.org/2.7/library/ssl.html#ssl.SSLSocket.do_handshake) during handshake. This flag is due on a SSLContext produced by create_default_context().

Note: when using tls_set() and tls_insecure_set() it works since in this case the SSLContext is created using ssl.SSLContext() and not create_default_context().

I think either tls_insecure_set() should update the check_hostname flag on the ssl context or add a note in documentation that tls_insecure_set should not be used with tls_set_context.


Step to reproduce:

```
# Certificate created with: openssl req -x509 -nodes -newkey rsa:1024 -keyout server.key -out server.pem -days 365 -subj '/CN=localhost' -batch
# mosquitto config:
# listener 1885
# cafile server.pem
# certfile server.pem
# keyfile server.key

import paho.mqtt.client
import ssl

cl2 = paho.mqtt.client.Client()
ssl_ctx = ssl.create_default_context(cafile='server.pem')
ssl_ctx.check_hostname = False  # Without this line, I does not work
cl2.tls_set_context(ssl_ctx)
cl2.tls_insecure_set(True)

# certificate was generated for CN=localhost, not "CN=127.0.0.1"
cl2.connect('127.0.0.1', 1885)
```

**Comment by Nzbuu** [comment]
@PierreF , that makes sense. I think it's just a matter of re-arranging some of the logic. I'll look at it in more detail later.

**Comment by Nzbuu** [comment]
OK. So I think it's pretty complicated to get all of the logic completely consistent. I'm going open an issue to propose removing SSL support in Python versions that don't include SSLContext (i.e. require Python 2.7.9+ or Python 3.2+). It would remove quite a lot of branches and make the intended behaviour clearer.

**Comment by ralight** [comment]
That sounds fine.

**Comment by Nzbuu** [comment]
@PierreF , your issue should be fixed now

File: src/paho/mqtt/client.py [hunk 1]
```diff
@@ -805,27 +826,27 @@ def reconnect(self):
...
+            # SSL is only supported when SSLContext is available (implies Python >= 2.7.9 or >= 3.2)
 
-        if self._transport == "websockets":
-            if self._tls_ca_certs is not None:
-                sock = WebsocketWrapper(sock, self._host, self._port, True)
+            verify_host = not self._tls_insecure
+            try:
+                # Try with server_hostname, even it's not supported in certain scenarios
+                sock = self._ssl_context.wrap_socket(sock, server_hostname=self._host)
+            except ValueError:
```
Comment (PierreF) [inline src/paho/mqtt/client.py]: ssl.CertificateError get caught by the except ValueError. Which means that if certificate validation failed (and server_hostname is supported) we will try to re-wrap the socket (which fail).

We probably should add a except ssl.CertificateError:\n raise juste before.

Comment (Nzbuu) [inline src/paho/mqtt/client.py]: :( Thanks for checking this out!

Comment (Nzbuu) [inline src/paho/mqtt/client.py]: Should be fixed now.

Looking at the ssl package docs, ssl.CertificateError should be the only error raised by the ssl code that isn't derived from OSError.

**Comment by Lokesh-K-Haralakatta** [comment]
When can we get the SNI support released with paho.mqtt.python pkg? We need this feature to support client side certificates in our connections to IoT Platform.

**Comment by Nzbuu** [comment]
@Lokesh-K-Haralakatta , might be better to ask at on the Paho-dev mailing list: https://dev.eclipse.org/mailman/listinfo/paho-dev

**Pull request merged**